File: python/leetcode/102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = {}

        q = deque()
        q.append((root, 0))

        while len(q) > 0:
            current, level = q.popleft()
            if current:
                if level in levels:
                    levels[level].append(current.val)
                else:
                    levels[level] = [current.val]

                if current.left:
                    q.append((current.left, level+1))
                if current.right:
                    q.append((current.right, level+1))

        return list(levels.values())

File: c#/leetcode/102-binary-tree-level-order-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> LevelOrder(TreeNode root) {
        if(root == null) {
            return new List<IList<int>>();
        }
        
        IList<IList<int>> res = new List<IList<int>>();
        Queue<TreeNode> q = new Queue<TreeNode>();
        
        q.Enqueue(root);
        
        while (q.Count > 0) {
            int count = q.Count;
            res.Add(new List<int>());
            
            for (int i = 0; i < count; i++) {
                TreeNode node = q.Dequeue();
                if(node.left != null) {
                    q.Enqueue(node.left);
                }
                if(node.right != null) {
                    q.Enqueue(node.right);
                }
                res.Last().Add(node.val);
            }
        }
        return res;
    }
}